File: Raven.Client.Lightweight/Connection/Profiling/ProfilingContext.cs
using Sparrow.Collections;
using System;
using System.Collections.Generic;

namespace Raven.Client.Connection.Profiling
{
    /// <summary>
    /// Manages all profiling activities for a given item
    /// </summary>
    public class ProfilingContext
    {
        private readonly ConcurrentLruSet<ProfilingInformation> leastRecentlyUsedCache = new ConcurrentLruSet<ProfilingInformation>(NumberOfSessionsToTrack);

        private const int NumberOfSessionsToTrack = 128;

        /// <summary>
        /// Register the action as associated with the sender
        /// </summary>
        public void RecordAction(object sender, RequestResultArgs requestResultArgs)
        {

            var profilingInformationHolder = sender as IHoldProfilingInformation;
            if (profilingInformationHolder == null)
                return;


            profilingInformationHolder.ProfilingInformation.Requests =
                new List<RequestResultArgs>(profilingInformationHolder.ProfilingInformation.Requests)
                {
                    requestResultArgs
                };

            leastRecentlyUsedCache.Push(profilingInformationHolder.ProfilingInformation);
        }

        /// <summary>
        /// Try to get a session matching the specified id.
        /// </summary>
        public ProfilingInformation TryGet(Guid id)
        {
            //TODO: verify that this doesn't break any tests
            //TODO: Maybe add more advanced predicates for when more than one request is sent per session.
            return leastRecentlyUsedCache.LastOrDefault(x => x.Id == id);
        }
    }
}

File: Raven.Client.Java/src/main/java/net/ravendb/client/connection/profiling/ProfilingContext.java
package net.ravendb.client.connection.profiling;

import java.util.UUID;

import net.ravendb.abstractions.closure.Function1;


public class ProfilingContext {
  private final ConcurrentLruSet<ProfilingInformation> leastRecentlyUsedCache = new ConcurrentLruSet<>(NUMBER_OF_SESSIONS_TO_TRACK, null);

  private static final int NUMBER_OF_SESSIONS_TO_TRACK = 128;

  /**
   * Register the action as associated with the sender
   */
  public void recordAction(Object sender, RequestResultArgs requestResultArgs) {

    if (sender instanceof IHoldProfilingInformation) {
      IHoldProfilingInformation profilingInformationHolder = (IHoldProfilingInformation) sender;
      profilingInformationHolder.getProfilingInformation().getRequests().add(requestResultArgs);
      leastRecentlyUsedCache.push(profilingInformationHolder.getProfilingInformation());
    }

  }

  /**
   * Try to get a session matching the specified id.
   * @param id
   * @return
   */
  public ProfilingInformation tryGet(final UUID id) {
    return leastRecentlyUsedCache.firstOrDefault(new Function1<ProfilingInformation, Boolean>() {

      @Override
      public Boolean apply(ProfilingInformation input) {
        return input.getId().equals(id);
      }
    });
  }
}
